vignettes/algorithm.Rmd

```diff
@@ -0,0 +1,341 @@
+we will proceed by example and use Gaussian univariate regression
+on the `trees` dataset. We will attempt to predict the volume of a tree based
+on girth and height.
+
+```{r}
+x <- with(trees, cbind(Girth, Height))
+y <- trees$Volume
+```
+
+### Standardization
+
+By default, the feature matrix in **sgdnet** is centered and scaled to unit
+variance. Here, however, we used the *biased* population standard deviation
+and divide by $n$ rather than $n-1$.
+
+```{r}
+sgdnet_sd <- function(x) sqrt(sum((x - mean(x))^2)/length(x))
+n <- nrow(x)
+x_bar <- colMeans(x)
+x_sd <- apply(x, 2, sgdnet_sd)
+x <- scale(x, center = x_bar, scale = x_sd)
+```
+
+In the univariate Gaussian case, we also standardize the response.
+
+```{r}
+y_bar <- mean(y)
+y_sd <- sgdnet_sd(y)
+y <- (y - y_bar)/y_sd
+```
+
+### Regularization path
+
+**sgdnet** supports $\ell_1$- and $\ell_2$-regularized regression as well
+as the elastic net, which is a linear combination of the two that is controlled
+by $\alpha$, such that
+
+$$\alpha = \frac{\lambda_2}{\lambda_1 + \lambda_2}$$
+
+where $\lambda_1$ and $\lambda_2$ denotes the amount of $\ell_1$ and $\ell_2$
+regularization respectively. 
+
+For our example, let's say that we are fitting the elastic net with
+$\alpha = 0.5$.
+
+Typically, we fit a model along a path of $\lambda$, beginning with 
+$\lambda_{max}$: the $\lambda$ at which the solution is completely sparse
+(save for the intercept, which is not regularized). It can be shown that
+
+$$\lambda_{\text{max}} = 
+  \max_{i}\frac{\langle\mathbf{x}_i, \mathbf{y}\rangle}{n\alpha}$$
+Of course, when $\alpha = 0$ as in ridge regression, this value becomes
+undefined and so we constrain it so that $\alpha \geq 0.001$ in the 
+computation of $\lambda_{\text{max}}$.
+
+For our example, we get the following.
+
+```{r}
+alpha <- 0.5
+(lambda_max <- max(abs(crossprod(y, x)))/(max(0.001, alpha)*n))
+```
+
+We then construct the $\lambda$ path to be a $\log_e$-spaced sequence starting
+at $\lambda_{\text{max}}$ and finishing at 
+$\lambda_{\text{max}} \times \lambda_{\text{min ratio}}$. Thus we have
+
+```{r}
+lambda.min.ratio <- 0.01
+nlambda <- 100
+lambda <- exp(seq(log(lambda_max),
+                  log(lambda_max*lambda.min.ratio),
+                  length.out = nlambda))
+head(lambda)
+```
+
+`lambda.min.ratio` and `nlambda` are both setable through the API, but 
+are given here at their defaults (per this example). Note that we return the
+lambda path on the original scale of the respone, $\mathbf{y}$, which means 
+that the user sees
+
+```{r}
+lambda_out <- lambda*y_sd
+head(lambda_out)
+```
+
+Naturally, whatever value the user provides in the argument `lambda` to
+`sgdnet()` will be scaled accordingly.
+
+### Step size
+
+The step size, $\gamma$, in SAGA is constant throughout the algorithm.
+For non-strongly convex objectives it is set to $1/(3L)$, where $L$ is the 
+*Lipschitz* constant. Since we are picking a single sample at a time,
```

Comment:
Be precise here - $L$ is the maximum Lipschitz constant of the term of the log-likelihood corresponding to a single observation, not the overall Lipschitz constant of the problem